File: ios-native/Leyne/Redesign/RedesignData.swift
// Mock data for the SG Transit redesign (iOS) — a faithful port of the design
// composition's constants. The redesign is a self-contained design surface, so
// it ships the same sample content the prototype used.

import SwiftUI

enum RDLoad { case seats, standing, packed }

struct RDArrival: Identifiable {
    let route: String
    let dest: String
    let load: RDLoad
    let min: String
    let then: String?
    // Identity is the service + destination (NOT the ETA) so a row persists as
    // its minutes tick down — letting the ETA `contentTransition` cross-fade
    // instead of the whole row being replaced.
    var id: String { route + dest }
}

struct RDStop: Identifiable {
    let name: String
    let code: String
    let dist: String
    let distShort: String
    let badge: String
    let arrivals: [RDArrival]
    var id: String { code }
}

struct RDStationDir: Identifiable {
    let to: String
    let via: String
    let plat: String
    let min: String
    let then: String
    var id: String { plat + to }
}

struct RDStation: Identifiable {
    let key: String
    let name: String
    let code: String
    let lineColor: Color
    let lineFg: Color
    let lineName: String
    let walk: String
    let freq: String
    let crowd: String
    let crowdLoad: RDLoad
    let firstTrain: String
    let lastTrain: String
    let exits: String
    let facilities: String
    let dirs: [RDStationDir]
    var id: String { key }
}

enum RDLineStatus { case normal, busy, major }

struct RDMrtLine: Identifiable {
    let code: String
    let name: String
    let badgeBg: Color
    let badgeFg: Color
    let statusText: String
    let location: String
    var major: Bool = false
    var detail: String? = nil
    var status: RDLineStatus = .normal
    var id: String { code }
}

private let orange = rdHex("FA9E0D")
private let onOrange = rdHex("3A2500")

let kRDStops: [RDStop] = [
    RDStop(
        name: "Opp Blk 123", code: "43091",
        dist: "You're at this stop · Farrer Road", distShort: "You are here", badge: "YOU'RE HERE",
        arrivals: [
            RDArrival(route: "165", dest: "HarbourFront Int", load: .seats, min: "2", then: "then 11"),
            RDArrival(route: "174", dest: "Clementi Int", load: .standing, min: "7", then: "then 16"),
            RDArrival(route: "186", dest: "Boon Lay Int", load: .packed, min: "12", then: nil),
            RDArrival(route: "5", dest: "Eunos Int", load: .seats, min: "4", then: "then 13"),
            RDArrival(route: "48", dest: "Marina Centre", load: .standing, min: "6", then: "then 18"),
            RDArrival(route: "93", dest: "Toa Payoh Int", load: .seats, min: "9", then: "then 21"),
            RDArrival(route: "961", dest: "Sin Ming Ave", load: .packed, min: "13", then: nil),
            RDArrival(route: "970", dest: "Jurong East Int", load: .standing, min: "16", then: nil),
        ]),
    RDStop(
        name: "Blk 240", code: "43099",
        dist: "1 min walk · 140 m", distShort: "140 m", badge: "1 MIN WALK",
        arrivals: [
            RDArrival(route: "51", dest: "Bishan Int", load: .seats, min: "5", then: "then 14"),
            RDArrival(route: "93", dest: "Toa Payoh", load: .standing, min: "8", then: nil),
            RDArrival(route: "410", dest: "Lor 1 Toa Payoh", load: .seats, min: "15", then: nil),
        ]),
    RDStop(
        name: "Farrer Rd Exit A", code: "43071",
        dist: "2 min walk · 160 m", distShort: "160 m", badge: "2 MIN WALK",
        arrivals: [
            RDArrival(route: "48", dest: "Marina Centre", load: .standing, min: "4", then: "then 12"),
            RDArrival(route: "93", dest: "Toa Payoh", load: .packed, min: "9", then: nil),
            RDArrival(route: "857", dest: "Yishun Int", load: .seats, min: "17", then: nil),
        ]),
]

let kRDStations: [String: RDStation] = [
    "holland": RDStation(
        key: "holland", name: "Holland Village", code: "CC21",
        lineColor: orange, lineFg: onOrange, lineName: "Circle Line",
        walk: "4 min walk · 320 m", freq: "every 4–6 min", crowd: "Moderate", crowdLoad: .standing,
        firstTrain: "5:31 AM", lastTrain: "12:18 AM", exits: "4 exits · A to D",
        facilities: "Lift, escalator & toilets",
        dirs: [
            RDStationDir(to: "HarbourFront", via: "via one-north · Buona Vista", plat: "A", min: "3", then: "then 9 min"),
            RDStationDir(to: "Dhoby Ghaut", via: "via Botanic Gardens · Bishan", plat: "B", min: "5", then: "then 12 min"),
        ]),
    "botanic": RDStation(
        key: "botanic", name: "Botanic Gardens", code: "CC19",
        lineColor: orange, lineFg: onOrange, lineName: "Circle Line · Downtown Line",
        walk: "7 min walk · 560 m", freq: "every 5–7 min", crowd: "Light", crowdLoad: .seats,
        firstTrain: "5:28 AM", lastTrain: "12:09 AM", exits: "3 exits · A to C",
        facilities: "Lift & escalator",
        dirs: [
            RDStationDir(to: "HarbourFront", via: "via Farrer Road · Holland V", plat: "A", min: "6", then: "then 13 min"),
            RDStationDir(to: "Marina Bay", via: "via Bishan · Promenade", plat: "B", min: "4", then: "then 11 min"),
        ]),
]

struct RDNearbyStation {
    let key: String
    let sub: String
    let topMin: String
}

let kRDNearbyStations: [RDNearbyStation] = [
    RDNearbyStation(key: "holland", sub: "4 min walk · Circle Line", topMin: "3"),
    RDNearbyStation(key: "botanic", sub: "7 min walk · CC · DT", topMin: "6"),
]

let kRDMrtLines: [RDMrtLine] = [
    RDMrtLine(code: "EWL", name: "East West Line", badgeBg: rdHex("009645"), badgeFg: .white,
              statusText: "MAJOR DELAY", location: "", major: true,
              detail: "Fault between Bugis and Tanah Merah. Add 15 min. Free bus bridging.", status: .major),
    RDMrtLine(code: "NSL", name: "North South Line", badgeBg: rdHex("D42E12"), badgeFg: .white,
              statusText: "Normal", location: "Bishan · 12 min", status: .normal),
    RDMrtLine(code: "CCL", name: "Circle Line", badgeBg: orange, badgeFg: onOrange,
              statusText: "Busy", location: "Farrer Road · 5 min", status: .busy),
    RDMrtLine(code: "NEL", name: "North East Line", badgeBg: rdHex("9900AA"), badgeFg: .white,
              statusText: "Normal", location: "Serangoon · 22 min", status: .normal),
]

File: lib/screens/redesign/redesign_data.dart
// Mock data for the SG Transit redesign — a faithful port of the constants in
// the design composition (__STOPS, __STATIONS and the MRT line list). The
// redesign is a self-contained design implementation, so it ships with the
// same sample content the prototype used rather than wiring the live LTA feed.

import 'package:flutter/widgets.dart';

/// Crowd / occupancy level, used for both bus loads and station crowding. Maps
/// to the bus (green) / amber / mrt (red) colour roles.
enum RdLoad { seats, standing, packed }

class RdArrival {
  const RdArrival({
    required this.route,
    required this.dest,
    required this.load,
    required this.min,
    this.then,
  });

  final String route;
  final String dest;
  final RdLoad load;
  final String min;
  final String? then;
}

class RdStop {
  const RdStop({
    required this.name,
    required this.code,
    required this.dist,
    required this.distShort,
    required this.badge,
    required this.arrivals,
  });

  final String name;
  final String code;
  final String dist;
  final String distShort;
  final String badge;
  final List<RdArrival> arrivals;
}

class RdStationDir {
  const RdStationDir({
    required this.to,
    required this.via,
    required this.plat,
    required this.min,
    required this.then,
  });

  final String to;
  final String via;
  final String plat; // single letter
  final String min;
  final String then;
}

class RdStation {
  const RdStation({
    required this.key,
    required this.name,
    required this.code,
    required this.lineColor,
    required this.lineFg,
    required this.lineName,
    required this.walk,
    required this.freq,
    required this.crowd,
    required this.crowdLoad,
    required this.firstTrain,
    required this.lastTrain,
    required this.exits,
    required this.facilities,
    required this.dirs,
  });

  final String key;
  final String name;
  final String code;
  final Color lineColor;
  final Color lineFg;
  final String lineName;
  final String walk;
  final String freq;
  final String crowd;
  final RdLoad crowdLoad;
  final String firstTrain;
  final String lastTrain;
  final String exits;
  final String facilities;
  final List<RdStationDir> dirs;
}

class RdMrtLine {
  const RdMrtLine({
    required this.code,
    required this.name,
    required this.badgeBg,
    required this.badgeFg,
    required this.statusText,
    required this.location,
    this.major = false,
    this.detail,
    this.status = RdLineStatus.normal,
  });

  final String code;
  final String name;
  final Color badgeBg;
  final Color badgeFg;
  final String statusText; // chip label for non-major lines
  final String location; // "Bishan · 12 min"
  final bool major;
  final String? detail; // body text for the major-delay card
  final RdLineStatus status;
}

enum RdLineStatus { normal, busy, major }

const _orange = Color(0xFFFA9E0D);
const _onOrange = Color(0xFF3A2500);

/// Bus stops, in priority order (index 0 is the user's current stop).
const List<RdStop> kRdStops = [
  RdStop(
    name: 'Opp Blk 123',
    code: '43091',
    dist: "You're at this stop · Farrer Road",
    distShort: 'You are here',
    badge: "YOU'RE HERE",
    arrivals: [
      RdArrival(route: '165', dest: 'HarbourFront Int', load: RdLoad.seats, min: '2', then: 'then 11'),
      RdArrival(route: '174', dest: 'Clementi Int', load: RdLoad.standing, min: '7', then: 'then 16'),
      RdArrival(route: '186', dest: 'Boon Lay Int', load: RdLoad.packed, min: '12'),
      RdArrival(route: '5', dest: 'Eunos Int', load: RdLoad.seats, min: '4', then: 'then 13'),
      RdArrival(route: '48', dest: 'Marina Centre', load: RdLoad.standing, min: '6', then: 'then 18'),
      RdArrival(route: '93', dest: 'Toa Payoh Int', load: RdLoad.seats, min: '9', then: 'then 21'),
      RdArrival(route: '961', dest: 'Sin Ming Ave', load: RdLoad.packed, min: '13'),
      RdArrival(route: '970', dest: 'Jurong East Int', load: RdLoad.standing, min: '16'),
    ],
  ),
  RdStop(
    name: 'Blk 240',
    code: '43099',
    dist: '1 min walk · 140 m',
    distShort: '140 m',
    badge: '1 MIN WALK',
    arrivals: [
      RdArrival(route: '51', dest: 'Bishan Int', load: RdLoad.seats, min: '5', then: 'then 14'),
      RdArrival(route: '93', dest: 'Toa Payoh', load: RdLoad.standing, min: '8'),
      RdArrival(route: '410', dest: 'Lor 1 Toa Payoh', load: RdLoad.seats, min: '15'),
    ],
  ),
  RdStop(
    name: 'Farrer Rd Exit A',
    code: '43071',
    dist: '2 min walk · 160 m',
    distShort: '160 m',
    badge: '2 MIN WALK',
    arrivals: [
      RdArrival(route: '48', dest: 'Marina Centre', load: RdLoad.standing, min: '4', then: 'then 12'),
      RdArrival(route: '93', dest: 'Toa Payoh', load: RdLoad.packed, min: '9'),
      RdArrival(route: '857', dest: 'Yishun Int', load: RdLoad.seats, min: '17'),
    ],
  ),
];

const Map<String, RdStation> kRdStations = {
  'holland': RdStation(
    key: 'holland',
    name: 'Holland Village',
    code: 'CC21',
    lineColor: _orange,
    lineFg: _onOrange,
    lineName: 'Circle Line',
    walk: '4 min walk · 320 m',
    freq: 'every 4–6 min',
    crowd: 'Moderate',
    crowdLoad: RdLoad.standing,
    firstTrain: '5:31 AM',
    lastTrain: '12:18 AM',
    exits: '4 exits · A to D',
    facilities: 'Lift, escalator & toilets',
    dirs: [
      RdStationDir(to: 'HarbourFront', via: 'via one-north · Buona Vista', plat: 'A', min: '3', then: 'then 9 min'),
      RdStationDir(to: 'Dhoby Ghaut', via: 'via Botanic Gardens · Bishan', plat: 'B', min: '5', then: 'then 12 min'),
    ],
  ),
  'botanic': RdStation(
    key: 'botanic',
    name: 'Botanic Gardens',
    code: 'CC19',
    lineColor: _orange,
    lineFg: _onOrange,
    lineName: 'Circle Line · Downtown Line',
    walk: '7 min walk · 560 m',
    freq: 'every 5–7 min',
    crowd: 'Light',
    crowdLoad: RdLoad.seats,
    firstTrain: '5:28 AM',
    lastTrain: '12:09 AM',
    exits: '3 exits · A to C',
    facilities: 'Lift & escalator',
    dirs: [
      RdStationDir(to: 'HarbourFront', via: 'via Farrer Road · Holland V', plat: 'A', min: '6', then: 'then 13 min'),
      RdStationDir(to: 'Marina Bay', via: 'via Bishan · Promenade', plat: 'B', min: '4', then: 'then 11 min'),
    ],
  ),
};

/// Nearby MRT stations shown on the home transfer card and the switch screen,
/// in order. The first entry is the "nearest station".
const List<({String key, String sub, String topMin})> kRdNearbyStations = [
  (key: 'holland', sub: '4 min walk · Circle Line', topMin: '3'),
  (key: 'botanic', sub: '7 min walk · CC · DT', topMin: '6'),
];

/// MRT & LRT line statuses for the Lines screen.
const List<RdMrtLine> kRdMrtLines = [
  RdMrtLine(
    code: 'EWL',
    name: 'East West Line',
    badgeBg: Color(0xFF009645),
    badgeFg: Color(0xFFFFFFFF),
    statusText: 'MAJOR DELAY',
    location: '',
    major: true,
    status: RdLineStatus.major,
    detail: 'Fault between Bugis and Tanah Merah. Add 15 min. Free bus bridging.',
  ),
  RdMrtLine(
    code: 'NSL',
    name: 'North South Line',
    badgeBg: Color(0xFFD42E12),
    badgeFg: Color(0xFFFFFFFF),
    statusText: 'Normal',
    location: 'Bishan · 12 min',
    status: RdLineStatus.normal,
  ),
  RdMrtLine(
    code: 'CCL',
    name: 'Circle Line',
    badgeBg: _orange,
    badgeFg: _onOrange,
    statusText: 'Busy',
    location: 'Farrer Road · 5 min',
    status: RdLineStatus.busy,
  ),
  RdMrtLine(
    code: 'NEL',
    name: 'North East Line',
    badgeBg: Color(0xFF9900AA),
    badgeFg: Color(0xFFFFFFFF),
    statusText: 'Normal',
    location: 'Serangoon · 22 min',
    status: RdLineStatus.normal,
  ),
];
